Source/ProgressionSystemRuntime/Private/Components/PSHUDComponent.cpp

```diff
@@ -0,0 +1,295 @@
+
+// Sets default values for this component's properties
+UPSHUDComponent::UPSHUDComponent()
+{
+	// Set this component to be initialized when the game starts, and to be ticked every frame.  You can turn these features
+	// off to improve performance if you don't need them.
+	PrimaryComponentTick.bCanEverTick = false;
+	PrimaryComponentTick.bStartWithTickEnabled = false;
+}
+
+// Called when the game starts
+void UPSHUDComponent::BeginPlay()
+{
+	Super::BeginPlay();
+
+	UE_LOG(LogProgressionSystem, Warning, TEXT("--- I'm running ---"));
+
+	AMyHUD* MyHUD = Cast<AMyHUD>(GetOwner());
+	const AMyHUD& HUD = *MyHUD;
+	ProgressionMenuWidgetInternal = HUD.CreateWidgetByClass<UPSMenuWidget>(UPSWorldSubsystem::Get().GetPSDataAsset()->GetProgressionMenuWidget(), true, 1);
+	ProgressionDataTableInternal = UPSWorldSubsystem::Get().GetPSDataAsset()->GetProgressionDataTable();
+
+	// Listen states to spawn widgets
+	if (AMyGameStateBase* MyGameState = UMyBlueprintFunctionLibrary::GetMyGameState())
+	{
+		HandleGameState(MyGameState);
+	}
+	else if (AMyPlayerController* MyPC = UMyBlueprintFunctionLibrary::GetLocalPlayerController())
+	{
+		MyPC->OnGameStateCreated.AddUniqueDynamic(this, &ThisClass::HandleGameState);
+	}
+	if (AMyPlayerState* MyPlayerState = UMyBlueprintFunctionLibrary::GetLocalPlayerState())
+	{
+		HandleEndGameState(MyPlayerState);
+	}
+	
+	UPSWorldSubsystem::Get().OnCurrentRowDataChanged.AddUObject(this, &ThisClass::OnPlayerTypeChanged);
+	// Save reference of this component to the world subsystem
+	UPSWorldSubsystem::Get().SetProgressionSystemComponent(this);
+	SaveGameInstanceInternal = UPSWorldSubsystem::Get().GetCurrentSaveGameData();
+	SavedProgressionRowDataInternal = UPSWorldSubsystem::Get().GetCurrentRowData();
+	
+	UpdateProgressionWidgetForPlayer();
+}
+
+void UPSHUDComponent::OnUnregister()
+{
+	Super::OnUnregister();
+	
+	if (ProgressionMenuWidgetInternal)
+	{
+		FWidgetUtilsLibrary::DestroyWidget(*ProgressionMenuWidgetInternal);
+		ProgressionMenuWidgetInternal = nullptr;
+	}
+	
+	ProgressionMenuWidgetInternal = nullptr;
+	ProgressionDataTableInternal = nullptr;
+}
+
+// Save the progression depends on EEndGameState
+void UPSHUDComponent::SavePoints(ELevelType Map, FPlayerTag Character, EEndGameState EndGameState)
+{
+	SavedProgressionRowDataInternal.CurrentLevelProgression += GetProgressionReward(Map, Character, EndGameState);
+
+	if (SavedProgressionRowDataInternal.CurrentLevelProgression >= SavedProgressionRowDataInternal.PointsToUnlock)
+	{
+		NextLevelProgressionRowData();
+	}
+	
+	SaveCurrentGameProgression();
+}
+
+// Returns rewards from data table for each type of game endings 
+int32 UPSHUDComponent::GetProgressionReward(ELevelType Map, FPlayerTag Character, EEndGameState EndGameState)
+{
+	FName CurrentProgressionRowName = GetProgressionRowName(Map, Character);
+	FPSRowData* ProgressionRowData = ProgressionDataTableInternal->FindRow<FPSRowData>(CurrentProgressionRowName, "Finding a needed row");
```

Comment:
Please don't cache Data Table anywhere and remove cache `ProgressionDataTableInternal` from HUD. Instead, in Subsystem, cache TMap once and then ask subsystem to return you rewards

Reply:
@JanSeliv actually, I do not need DataTable. It's used only for the init of the save file and not used anymore. So, removed the DataTable from HUD, migrated geting rewards to subsystem. Done.

Reply:
I still see you have `ProgressionDataTableInternal` in the `PSWorldSubsystem`:
1. Remove cached `TObjectPtr<const UDataTable> ProgressionDataTableInternal` pointer to this data table from `PSWorldSubsystem`, keep its UDataTable pointer only in the Data Asset, don't populate\cache outside.
2. (optional) If you fill you would need access to data table more than once (so you want to cache it), then in `PSWorldSubsystem` cache `TMap<FName, FPSRowData>` rows directly instead of `TObjectPtr<const UDataTable>` pointer

Reply:
done